pandas/core/generic.py

```diff
@@ -2012,6 +2013,15 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
         # Check weights for compliance
         if weights is not None:
 
+            # If a series, align with frame
+            if isinstance(weights, pd.Series):
+                obj_index = self.axes[axis]
+
+                if len(obj_index.difference(weights.index)) != 0:
```

Comment:
@jreback My primary motivation for this PR is ensuring the weights and sampled object are aligned on their indices (since the actual sampling is done using positions, not indexes, this has to be done separately). 

My concern with a full "re-index" that drops labels in the weight `Series` that aren't in the sampled object is that it could obscure user errors. For example, if someone screwed up an index and passed the following, re-indexing would drop the row with the (possibly mis-typed) weight silently and hide the mistake:

    df = DataFrame({'col1':[5,6,7], 'col2':['a','b','c'], }, index = [9,5,3])
    s = Series([0.7,0,0.3], index=[3,5,10])
    
    df.sample(n=1, weight=s)

It seems like bad defensive programming to let that pass, especially when it's hard for me to think of a reason that someone would want to pass weights with extra labels (and if they did, I don't mind requiring them to deal with that outside the function). 

If you think it's important to allow for that, I would prefer to add a new option to the function so users have to intentionally enable a full re-indexing.